neuralmonkey/decoders/sequence_decoder.py

```diff
@@ -0,0 +1,258 @@
+            return tf.get_variable(
+                "logit_bias", [len(self.vocabulary)],
+                initializer=tf.constant_initializer(
+                    - math.log(len(self.vocabulary))))
+
+    def get_logits(self, state: tf.Tensor) -> tf.Tensor:
+        state = dropout(state, self.dropout_keep_prob, self.train_mode)
+        return tf.matmul(state, self.decoding_w) + self.decoding_b
+
+    @tensor
+    def train_logits(self) -> tf.Tensor:
+        # THE LAST TRAIN INPUT IS NOT USED IN DECODING FUNCTION
+        # (just as a target)
+        logits, _, _ = self.decoding_loop(train_mode=True)
+        return logits
+
+    @tensor
+    def train_logprobs(self) -> tf.Tensor:
+        return tf.nn.log_softmax(self.train_logits)
+
+    @tensor
+    def train_xents(self) -> tf.Tensor:
+        train_targets = tf.transpose(self.train_inputs)
+
+        return tf.contrib.seq2seq.sequence_loss(
+            tf.transpose(self.train_logits, perm=[1, 0, 2]),
+            train_targets,
+            tf.transpose(self.train_mask),
+            average_across_batch=False)
+
+    @tensor
+    def train_loss(self) -> tf.Tensor:
+        return tf.reduce_mean(self.train_xents)
+
+    @property
+    def cost(self) -> tf.Tensor:
+        return self.train_loss
+
+    @tensor
+    def runtime_loop_result(self) -> Tuple[tf.Tensor, tf.Tensor, tf.Tensor]:
+        return self.decoding_loop(train_mode=False)
+
+    @tensor
+    def runtime_logits(self) -> tf.Tensor:
+        return tuple(self.runtime_loop_result)[0]
+
+    @tensor
+    def runtime_rnn_states(self) -> tf.Tensor:
+        return tuple(self.runtime_loop_result)[1]
+
+    @tensor
+    def runtime_mask(self) -> tf.Tensor:
+        return tuple(self.runtime_loop_result)[2]
+
+    @tensor
+    def decoded(self) -> tf.Tensor:
+        return tf.argmax(self.runtime_logits[:, :, 1:], -1) + 1
+
+    @tensor
+    def runtime_loss(self) -> tf.Tensor:
+        train_targets = tf.transpose(self.train_inputs)
+        batch_major_logits = tf.transpose(self.runtime_logits, [1, 0, 2])
+        min_time = tf.minimum(tf.shape(train_targets)[1],
+                              tf.shape(batch_major_logits)[1])
+
+        # TODO if done properly, there should be padding of the shorter
+        # sequence instead of cropping to the length of the shorter one
+
+        return tf.contrib.seq2seq.sequence_loss(
+            logits=batch_major_logits[:, :min_time],
+            targets=train_targets[:, :min_time],
+            weights=tf.transpose(self.train_mask)[:, :min_time])
+
+    @tensor
+    def runtime_logprobs(self) -> tf.Tensor:
+        return tf.nn.log_softmax(self.runtime_logits)
+
+    @property
+    def output_dimension(self) -> Union[int, tf.Tensor]:
+        raise NotImplementedError("Abstract property")
+
+    def get_initial_loop_state(self) -> LoopState:
+        raise NotImplementedError("Abstract method")
+
+    def loop_continue_criterion(self, *args) -> tf.Tensor:
```

Comment:
docstring
Decide whether to break out of the while loop

Args: LoopState

Reply:
DONE